### PR TITLE
Feature/ci prevent new cppcheck issues

### PR DESCRIPTION
**Pull request by rmelotte**
With the huge amount of issues cppcheck reports already it will take a lot of time to solve them all, and it could be useful to still prevent new issues like we do for PEP8.

Since we've never been block by "less severe" cppcheck issues before, this means we will potentially have to tweak cppcheck from time to time for issues we think are not relevant, and PR might get block by it in the meantime.

### CODE REVIEW

**Comment by rmelotte** [comment]
~~We could also make the job "allowed to fail" for some time, until we are confident it only reports relevant issues.~~
As @arnout mentioned during the sync meeting, making the job "allowed to fail" will encourage people to **not** check its status, which means it won't be taken into account in the end.

**Review by tomereli — APPROVED** [review]

File: ci/cppcheck/cppcheck-new-issues.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+# shellcheck source=../../tools/function.sh
+. "${rootdir}/tools/functions.sh"
+
+
+if [ ! -f "$rootdir/cppcheck_results.txt" ] ; then
+    err "run $rootdir/tools/docker/static-analysis/cppcheck.sh first"
+fi
+
+mapfile -t issues < <(grep -E ".+:[0-9]+:[0-9]+: .+: .+ \[.+\]$"  "$rootdir/cppcheck_results.txt")
```
Comment (arnout) [inline ci/cppcheck/cppcheck-new-issues.sh]: TIL: the mapfile command. Interesting!

File: ci/cppcheck/cppcheck-new-issues.sh [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+scriptdir="$(cd "${0%/*}" && pwd)"
+rootdir="${scriptdir%/*/*}"
+
+# shellcheck source=../../tools/function.sh
+. "${rootdir}/tools/functions.sh"
+
+
+if [ ! -f "$rootdir/cppcheck_results.txt" ] ; then
+    err "run $rootdir/tools/docker/static-analysis/cppcheck.sh first"
```
Comment (arnout) [inline ci/cppcheck/cppcheck-new-issues.sh]: Missing exit after printing the error.

File: ci/cppcheck/cppcheck-new-issues.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+fi
+
+mapfile -t issues < <(grep -E ".+:[0-9]+:[0-9]+: .+: .+ \[.+\]$"  "$rootdir/cppcheck_results.txt")
+
+status=0
+for i in "${issues[@]}"
```
Comment (arnout) [inline ci/cppcheck/cppcheck-new-issues.sh:16]: (nitpick) As happy as I am with learning the mapfile command, here it really doesn't bring anything over the traditional (and POSIX)

```
grep -E ... | \
    while read -r i; do
```

**Review by arnout — APPROVED** [review]
I will approve this PR, but I'm afraid we're going to revert it pretty quickly...

The problem is that the issue matching happens based on the line numbers. For example, if we add an include to db.cpp, all the errors that it currently gives will shift by one line. So we need to add an 800-line diff to cppcheck_existing_issues.txt to handle that... The result will be that people blindly overwrite cppcheck_existing_issues.txt so you still don't see a new issue.

To work around it, we could use a different output format where the code is on the same line as the rest, and where we don't have line and column numbers.

For now though, I'd merge as is. We can still fix it if it turns out to be an actual problem.

**Comment by rmelotte** [comment]
Indeed the line number is a pretty annoying issue, I'm glad you mentioned it.
In the end it was easy to change the format of the reported issues since we're doing this in a separate job anyway.
So I did exactly what you suggested: removed line and column numbers and added the code itself.

**Pull request merged**